File: linear_time_sorting/CountingSort/CountingSort.py
def counting_sort(arr):
    if len(arr) == 0:
        return arr

    # Find the maximum and minimum elements in the array
    max_val = max(arr)
    min_val = min(arr)
    range_of_elements = max_val - min_val + 1

    # Create a count array to store the count of each unique element
    count = [0] * range_of_elements
    output = [0] * len(arr)

    # Store the count of each element
    for num in arr:
        count[num - min_val] += 1

    # Modify the count array by adding the count of the previous element
    for i in range(1, range_of_elements):
        count[i] += count[i - 1]

    # Build the output array
    for num in reversed(arr):
        output[count[num - min_val] - 1] = num
        count[num - min_val] -= 1

    # Copy the sorted elements back to the original array
    for i in range(len(arr)):
        arr[i] = output[i]

    return arr

# Example usage
if __name__ == "__main__":
    arr = [4, 2, 2, 8, 3, 3, 1]
    print("Given array is:", arr)
    sorted_arr = counting_sort(arr)
    print("Sorted array is:", sorted_arr)

File: linear_time_sorting/CountingSort/CountingSort.go
package main

import (
    "fmt"
)

// countingSort sorts an array using the Counting Sort algorithm
func countingSort(arr []int) []int {
    if len(arr) == 0 {
        return arr
    }

    // Find the maximum and minimum elements in the array
    maxVal := arr[0]
    minVal := arr[0]
    for _, num := range arr {
        if num > maxVal {
            maxVal = num
        }
        if num < minVal {
            minVal = num
        }
    }
    rangeOfElements := maxVal - minVal + 1

    // Create a count array to store the count of each unique element
    count := make([]int, rangeOfElements)
    output := make([]int, len(arr))

    // Store the count of each element
    for _, num := range arr {
        count[num-minVal]++
    }

    // Modify the count array by adding the count of the previous element
    for i := 1; i < rangeOfElements; i++ {
        count[i] += count[i-1]
    }

    // Build the output array
    for i := len(arr) - 1; i >= 0; i-- {
        output[count[arr[i]-minVal]-1] = arr[i]
        count[arr[i]-minVal]--
    }

    // Copy the sorted elements back to the original array
    for i := 0; i < len(arr); i++ {
        arr[i] = output[i]
    }

    return arr
}

func main() {
    arr := []int{4, 2, 2, 8, 3, 3, 1}
    fmt.Println("Given array is:", arr)
    sortedArr := countingSort(arr)
    fmt.Println("Sorted array is:", sortedArr)
}